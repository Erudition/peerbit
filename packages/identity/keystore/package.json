{
  "name": "@dao-xyz/orbit-db-keystore",
  "version": "0.0.1",
  "description": "A local key manager for OrbitDB",
  "sideEffects": false,
  "main": "lib/cjs/index.js",
  "module": "lib/esm/index.mjs",
  "types": "lib/esm/index.d.ts",
  "exports": {
    "import": "./lib/esm/index.mjs",
    "require": "./lib/cjs/index.js"
  },
  "files": [
    "lib",
    "src",
    "LICENSE"
  ],
  "publishConfig": {
    "access": "public"
  },
  "scripts": {
    "clean": "shx rm -rf lib/*",
    "build": "yarn clean && tsc -p tsconfig.json; tsc-esm -p tsconfig.json && tsc -p tsconfig.cjs.json",
    "postbuild": "echo '{\"type\":\"commonjs\"}' | node ../../../node_modules/.bin/json > lib/cjs/package.json && echo '{\"type\":\"module\"} ' | node ../../../node_modules/.bin/json > lib/esm/package.json",
    "test": "node ../../../node_modules/.bin/jest test  -c  ../../../jest.config.ts --runInBand --forceExit",
    "test:unit": "node ../../../node_modules/.bin/jest test  -c  ../../../jest.config.unit.ts --runInBand --forceExit",
    "test:integration": "node ../node_modules/.bin/jest test -c  ../../../jest.config.integration.ts --runInBand --forceExit"
  },
  "author": "Haad",
  "license": "MIT",
  "dependencies": {
    "@dao-xyz/borsh": "^3.2.5",
    "@dao-xyz/borsh-utils": "^0.0.1",
    "@dao-xyz/encryption-utils": "^0.0.1",
    "elliptic": "^6.5.4",
    "level": "^8.0.0",
    "libp2p-crypto": "^0.19.4",
    "lru-cache": "^7.14.0",
    "mkdirp": "^0.5.5",
    "reachdown": "^1.1.0",
    "safe-buffer": "^5.2.1",
    "sodium-plus": "^0.9.0",
    "tweetnacl": "^1.0.3",
    "@dao-xyz/time": "^0.0.11"
  },
  "devDependencies": {
    "@babel/core": "^7.10.5",
    "@babel/plugin-syntax-object-rest-spread": "^7.8.3",
    "@babel/plugin-transform-modules-commonjs": "^7.10.4",
    "@babel/plugin-transform-runtime": "^7.10.5",
    "@babel/preset-env": "^7.10.4",
    "@babel/runtime": "^7.10.5",
    "@types/libsodium-wrappers": "^0.7.9",
    "babel-loader": "~8.0.4",
    "cross-env": "^7.0.2",
    "fruitdown": "^1.0.2",
    "is-node": "^1.0.2",
    "json-loader": "~0.5.7",
    "jsondown": "^1.0.0",
    "localstorage-down": "^0.6.7",
    "memdown": "^6.0.0",
    "mocha-headless-chrome": "^2.0.3",
    "@dao-xyz/orbit-db-test-utils": "^0.0.1",
    "standard": "^12.0.1",
    "webpack": "^4.43.0",
    "webpack-cli": "^3.3.12"
  },
  "bugs": "https://github.com/orbitdb/orbit-db-keystore/issues",
  "homepage": "https://github.com/orbitdb/orbit-db-keystore",
  "contributors": [
    "adam-palazzo",
    "haadcode",
    "RichardLitt",
    "shamb0t",
    "thiagodelgado111",
    "willemneal"
  ],
  "standard": {
    "env": [
      "mocha"
    ]
  },
  "localMaintainers": [
    "haad <haad@haja.io>",
    "shamb0t <shams@haja.io>",
    "hajamark <mark@haja.io>"
  ]
}