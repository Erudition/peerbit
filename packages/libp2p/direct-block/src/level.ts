import { BlockStore, PutOptions } from "./store.js";
import {
	cidifyString,
	codecCodes,
	defaultHasher,
	stringifyCid,
} from "./block.js";
import * as Block from "multiformats/block";
import { AbstractBatchOperation, AbstractLevel } from "abstract-level";
import { MemoryLevel } from "memory-level";
import { waitFor } from "@dao-xyz/peerbit-time";
import { ByteView } from "@ipld/dag-cbor";

export type LevelBatchOptions = {
	interval?: number;
	onError?: (error: any) => void;
};
export type LevelBlockStoreOptions = { batch: LevelBatchOptions | true };
export class LevelBlockStore implements BlockStore {
	_level: AbstractLevel<any, string, Uint8Array>;
	_opening: Promise<any>;
	_closed = false;
	_onClose: (() => any) | undefined;
	_batchOptions?: { interval: number; onError?: (e: any) => void };
	_interval: any;
	_txQueue: AbstractBatchOperation<
		AbstractLevel<any, string, Uint8Array>,
		string,
		Uint8Array
	>[];
	_tempStore: Map<string, ByteView<any>>;
	_txPromise: Promise<void>;

	constructor(
		level: AbstractLevel<any, string, Uint8Array>,
		options?: LevelBlockStoreOptions
	) {
		this._level = level;
		if (options?.batch) {
			this._batchOptions =
				typeof options.batch === "boolean"
					? {
							interval: 300,
					  }
					: {
							interval: options.batch.interval || 100,
							onError: options.batch.onError,
					  };
			if (this._batchOptions!.interval <= 0) {
				throw new Error(
					"Batch interval needs to be greater than 0 or undefined"
				);
			}
		}
	}

	async get<T>(
		cid: string,
		options?: {
			raw?: boolean;
			links?: string[];
			timeout?: number;
			hasher?: any;
		}
	): Promise<Block.Block<T, any, any, any> | undefined> {
		const cidObject = cidifyString(cid);
		try {
			const bytes =
				(this._tempStore && this._tempStore.get(cid)) ||
				(await this._level.get(cid, { valueEncoding: "view" }));
			if (!bytes) {
				return undefined;
			}
			const codec = codecCodes[cidObject.code];
			const block = await Block.decode({
				bytes,
				codec,
				hasher: options?.hasher || defaultHasher,
			});
			return block as Block.Block<T, any, any, any>;
		} catch (error: any) {
			if (
				typeof error?.code === "string" &&
				error?.code?.indexOf("LEVEL_NOT_FOUND") !== -1
			) {
				return undefined;
			}
			throw error;
		}
	}

	async put<T>(
		block: Block.Block<T, any, any, any>,
		options?: PutOptions
	): Promise<string> {
		const cid = stringifyCid(block.cid);
		const bytes = block.bytes;

		if (this._batchOptions) {
			this._tempStore.set(cid, bytes);
			this._txQueue.push({
				type: "put",
				key: cid,
				value: bytes,
				valueEncoding: "view",
			});
		} else {
			await this._level.put(cid, bytes, {
				valueEncoding: "view",
			});
		}

		return cid;
	}

	async rm(cid: string): Promise<void> {
		if (this._batchOptions) {
			this._txQueue.push({ type: "del", key: cid });
		} else {
			await this._level.del(cid);
		}
	}

	async open(): Promise<void> {
		this._closed = false;
		if (this._level.status !== "opening" && this._level.status !== "open") {
			await this._level.open();
		}

		if (this._batchOptions) {
			this._txQueue = [];
			this._tempStore = new Map();

			this._interval = setInterval(() => {
				if (this._level.status === "open" && this._txQueue.length > 0) {
					try {
						const arr = this._txQueue.splice(0, this._txQueue.length);
						if (arr?.length > 0) {
							this._txPromise = (
								this._txPromise ? this._txPromise : Promise.resolve()
							).finally(() => {
								return this._level.batch(arr).then(() => {
									arr.forEach((v) => {
										if (v.type === "put") {
											this._tempStore.delete(v.key);
										} else if (v.type === "del") {
											this._tempStore.delete(v.key);
										}
									});
								});
							});
						}
					} catch (error) {
						this._batchOptions?.onError && this._batchOptions.onError(error);
					}
				}
			}, this._batchOptions.interval);
		}

		try {
			this._opening = waitFor(() => this._level.status === "open", {
				delayInterval: 100,
				timeout: 10 * 1000,
				stopperCallback: (fn) => {
					this._onClose = fn;
				},
			});
			await this._opening;
		} catch (error) {
			if (this._closed) {
				return;
			}
			throw error;
		} finally {
			this._onClose = undefined;
		}
	}

	async close(): Promise<void> {
		await this.idle();
		if (this._batchOptions) {
			clearInterval(this._interval);
			this._interval = undefined;
			this._tempStore.clear();
		}
		this._closed = true;
		this._onClose && this._onClose();
		return this._level.close();
	}

	async idle(): Promise<void> {
		if (this._txQueue) {
			await waitFor(() => this._txQueue.length === 0);
			await this._txPromise;
		}
	}

	get status() {
		if (this._batchOptions && !this._interval) {
			return "closed";
		}
		return this._level.status;
	}
}

export class MemoryLevelBlockStore extends LevelBlockStore {
	constructor(options?: LevelBlockStoreOptions) {
		super(new MemoryLevel({ valueEncoding: "view" }), options);
	}
}
