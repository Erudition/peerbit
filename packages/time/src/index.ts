export const delay = (ms: number, stopperCallback?: (stopper: () => void) => void) => {
    return new Promise((res, rej) => {
        if (stopperCallback)
            stopperCallback(() => res(true));
        setTimeout(res, ms)
    })
};


export const waitFor = async (fn: () => boolean | Promise<boolean>, timeout: number = 10 * 1000, stopperCallback?: (stopper: () => void) => void) => {

    let startTime = +new Date;
    let stop = false
    if (stopperCallback) {
        const stopper = () => { stop = true }
        stopperCallback(stopper);
    }
    while (+new Date - startTime < timeout) {
        if (stop) {
            return;
        }
        if (await fn()) {
            return;
        }
        await delay(50);

    }
    throw new Error("Timed out")

};
export const waitForAsync = async (fn: () => Promise<boolean>, timeout: number = 10 * 1000, stopperCallback?: (stopper: () => void) => void) => {

    let startTime = +new Date;
    let stop = false
    if (stopperCallback) {
        const stopper = () => { stop = true }
        stopperCallback(stopper);
    }
    while (+new Date - startTime < timeout) {
        if (stop) {
            return;
        }
        if (await fn()) {
            return;
        }
        await delay(50);
    }
    throw new Error("Timed out")
};
