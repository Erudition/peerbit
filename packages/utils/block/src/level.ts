import { BlockStore, PutOptions } from "./store.js";
import {
    cidifyString,
    codecCodes,
    defaultHasher,
    stringifyCid,
} from "./block.js";
import * as Block from "multiformats/block";
import { AbstractLevel } from "abstract-level";
import { MemoryLevel } from "memory-level";
import { waitFor } from "@dao-xyz/peerbit-time";

abstract class LevelBlockStore implements BlockStore {
    _level: AbstractLevel<any, string, Uint8Array>;
    _opening: Promise<any>;
    _closed = false;
    _onClose: (() => any) | undefined;
    constructor(level: AbstractLevel<any, string, Uint8Array>) {
        this._level = level;
    }

    async get<T>(
        cid: string,
        options?: {
            raw?: boolean;
            links?: string[];
            timeout?: number;
            hasher?: any;
        }
    ): Promise<Block.Block<T, any, any, any> | undefined> {
        const cidObject = cidifyString(cid);
        try {
            const bytes = await this._level.get(cid);
            const codec = codecCodes[cidObject.code];
            const block = await Block.decode({
                bytes,
                codec,
                hasher: options?.hasher || defaultHasher,
            });
            return block as Block.Block<T, any, any, any>;
        } catch (error: any) {
            if (
                typeof error?.code === "string" &&
                error?.code?.indexOf("LEVEL_NOT_FOUND") !== -1
            ) {
                return undefined;
            }
            throw error;
        }
    }

    async put(
        block: Block.Block<any, any, any, any>,
        options?: PutOptions
    ): Promise<string> {
        await this._level.put(stringifyCid(block.cid), block.bytes, {
            valueEncoding: "view",
        });
        return stringifyCid(block.cid);
    }

    async rm(cid: string): Promise<void> {
        await this._level.del(cid);
    }

    async open(): Promise<void> {
        this._closed = false;
        if (this._level.status !== "opening" && this._level.status !== "open") {
            await this._level.open();
        }
        try {
            this._opening = waitFor(() => this._level.status === "open", {
                delayInterval: 100,
                timeout: 10 * 1000,
                stopperCallback: (fn) => {
                    this._onClose = fn;
                },
            });
            await this._opening;
        } catch (error) {
            if (this._closed) {
                return;
            }
            throw error;
        } finally {
            this._onClose = undefined;
        }
    }

    async close(): Promise<void> {
        this._closed = true;
        this._onClose && this._onClose();
        return this._level.close();
    }
}

export class MemoryLevelBlockStore extends LevelBlockStore {
    constructor() {
        super(new MemoryLevel({ valueEncoding: "view" }));
    }
}
