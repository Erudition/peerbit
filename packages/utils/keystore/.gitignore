node_modules/
keystore/
keystore-test/
test/browser/bundle.js
test/browser/bundle.js.map
examples/browser/peerbit-keystore.min.js*
dist/*.js.map
