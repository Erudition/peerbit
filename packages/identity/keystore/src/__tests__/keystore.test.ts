
import path from 'path';
import assert from 'assert';
import LRU from 'lru-cache';
import { BoxKeyWithMeta, createStore, Keystore, KeyWithMeta, SignKeyWithMeta } from '../keystore';
import rmrf from 'rimraf'
import { waitFor } from '@dao-xyz/time';
import { Level } from 'level';
import { Ed25519PublicKey, X25519PublicKey, X25519SecretKey } from 'sodium-plus';

import fs from 'fs-extra'
let store: Level;
const fixturePath = path.join('packages/identity/orbit-db-keystore/src/__tests__', 'fixtures', 'signingKeys')
const storagePath = path.join('packages/identity/orbit-db-keystore/src/__tests__', 'signingKeys')
const upgradePath = path.join('packages/identity/orbit-db-keystore/src/__tests__', 'upgrade')
const tempKeyPath = "packages/identity/orbit-db-keystore/src/__tests__/keystore-test";
jest.useRealTimers();
jest.setTimeout(100000);

describe('keystore', () => {

  beforeAll(async () => {
    await fs.copy(fixturePath, storagePath)
    rmrf.sync(tempKeyPath)
    store = await createStore(tempKeyPath) // storagePath

  })

  afterAll(async () => {
    rmrf.sync(storagePath)
    rmrf.sync(upgradePath)
    rmrf.sync(tempKeyPath)

  })

  describe('constructor', () => {
    it('creates a new Keystore instance', async () => {
      const keystore = new Keystore(store)

      expect(typeof keystore.close).toEqual('function')
      expect(typeof keystore.openStore).toEqual('function')
      expect(typeof keystore.hasKey).toEqual('function')
      expect(typeof keystore.createKey).toEqual('function')
      expect(typeof keystore.getKeyByPath).toEqual('function')
    })

    it('assigns this._store', async () => {
      const keystore = new Keystore(store)
      // Loose check for leveldownishness
      assert(['open', 'opening'].includes(keystore._store.status))
    })

    it('assigns this.cache with default of 100', async () => {
      const keystore = new Keystore(store)
      expect(keystore._cache.max).toEqual(100)
    })

    it('creates a proper leveldown / level-js store if not passed a store', async () => {
      const keystore = new Keystore()
      expect(keystore._store.status).toEqual('opening')
      await keystore.close()
    })

    it('creates a keystore with empty options', async () => {
      const keystore = new Keystore({})
      expect(keystore._store.status).toEqual('opening')
      await keystore.close()
    })

    it('creates a keystore with only cache', async () => {
      const cache = new LRU({ max: 10 })
      const keystore = new Keystore({ cache })
      expect(keystore._store.status).toEqual('opening')
      assert(keystore._cache === cache)
      await keystore.close()
    })

    it('creates a keystore with both', async () => {
      const cache = new LRU({ max: 10 })
      const keystore = new Keystore({ store, cache })
      assert(['open', 'opening'].includes(keystore._store.status))
      assert(keystore._cache === cache)
      assert(keystore._store === store)
    })
  })


  describe('createKey', () => {
    let keystore: Keystore

    beforeEach(async () => {
      keystore = new Keystore(store)
      if (store.status !== 'open') {
        await store.open()
      }
    })

    it('creates a new key', async () => {
      const id = 'a new key'
      await keystore.createKey(id, SignKeyWithMeta)
      const hasKey = await keystore.hasKey(id, SignKeyWithMeta)
      expect(hasKey).toEqual(true)
    })

    it('throws an error upon not receiving an ID', async () => {
      try {
        await keystore.createKey(undefined, undefined)
      } catch (e) {
        assert(true)
      }
    })
    it('throws an error if key already exist', async () => {
      const id = 'already'
      await keystore.createKey(id, SignKeyWithMeta)
      try {
        await keystore.createKey(id, SignKeyWithMeta)
      } catch (e) {
        assert(true)
      }
    })
    it('throws an error accessing a closed store', async () => {
      try {
        const id = 'X'

        await store.close()
        await keystore.createKey(id, SignKeyWithMeta)
      } catch (e) {
        assert(true)
      }
    })



    afterEach(async () => {
      // await keystore.close()
    })
  })

  describe('saveKey', () => {
    let keystore: Keystore

    beforeEach(async () => {
      keystore = new Keystore(store)
      if (store.status !== 'open') {
        await store.open()
      }
    })

    it('can overwrite if secret key is missing', async () => {
      const id = 'overwrite key'
      let keyWithMeta = new BoxKeyWithMeta({
        secretKey: undefined,
        publicKey: new X25519PublicKey(Buffer.from(new Array(32).fill(0))),
        timestamp: BigInt(+new Date),
        group: '_'
      });
      let savedKey = await keystore.saveKey(keyWithMeta, id)
      assert(!savedKey.secretKey);
      keyWithMeta = new BoxKeyWithMeta({
        secretKey: new X25519SecretKey(Buffer.from(new Array(32).fill(0))),
        publicKey: new X25519PublicKey(Buffer.from(new Array(32).fill(0))),
        timestamp: keyWithMeta.timestamp,
        group: '_'
      });
      savedKey = await keystore.saveKey(keyWithMeta, id)
      assert(!!savedKey.secretKey)
    })

    it('will return secret key if missing when saving', async () => {
      const id = 'overwrite key 2'
      let keyWithMeta = new BoxKeyWithMeta({
        secretKey: new X25519SecretKey(Buffer.from(new Array(32).fill(0))),
        publicKey: new X25519PublicKey(Buffer.from(new Array(32).fill(0))),
        timestamp: BigInt(+new Date),
        group: '_'
      });
      let savedKey = await keystore.saveKey(keyWithMeta, id)
      assert(!!savedKey.secretKey);
      keyWithMeta = new BoxKeyWithMeta({
        secretKey: undefined,
        publicKey: new X25519PublicKey(Buffer.from(new Array(32).fill(0))),
        timestamp: keyWithMeta.timestamp,
        group: '_'
      });
      savedKey = await keystore.saveKey(keyWithMeta, id)
      assert(!!savedKey.secretKey)
    })

    it('throws an error upon not receiving an ID', async () => {
      try {
        await keystore.createKey(undefined, undefined)
      } catch (e) {
        assert(true)
      }
    })
    it('throws an error if key already exist', async () => {
      const id = 'already'
      await keystore.createKey(id, SignKeyWithMeta)
      try {
        await keystore.createKey(id, SignKeyWithMeta)
      } catch (e) {
        assert(true)
      }
    })
    it('throws an error accessing a closed store', async () => {
      try {
        const id = 'X'

        await store.close()
        await keystore.createKey(id, SignKeyWithMeta)
      } catch (e) {
        assert(true)
      }
    })



    afterEach(async () => {
      // await keystore.close()
    })
  })

  describe('hasKey', () => {
    let keystore: Keystore

    beforeAll(async () => {
      if (store.status !== 'open') {
        await store.open()
      }
      keystore = new Keystore(store)
      await keystore.createKey('YYZ', SignKeyWithMeta)
    })

    it('returns true if key exists', async () => {
      const hasKey = await keystore.hasKey('YYZ', SignKeyWithMeta)
      expect(hasKey).toEqual(true)
    })

    it('returns false if key does not exist', async () => {
      let hasKey
      try {
        hasKey = await keystore.hasKey('XXX', SignKeyWithMeta)
      } catch (e) {
        expect(hasKey).toEqual(true)
      }
    })

    it('throws an error upon not receiving an ID', async () => {
      try {
        await keystore.hasKey(undefined, undefined)
      } catch (e) {
        assert(true)
      }
    })

    it('throws an error accessing a closed store', async () => {
      try {
        await store.close()
        await keystore.hasKey('XXX', SignKeyWithMeta)
      } catch (e) {
        assert(true)
      }
    })

    afterEach(async () => {
      // await keystore.close()
    })
  })

  describe('getKey', () => {
    let keystore: Keystore, createdKey: SignKeyWithMeta
    beforeAll(async () => {
      if (store.status !== 'open') {
        await store.open()
      }
      keystore = new Keystore(store)
      createdKey = await keystore.createKey('ZZZ', SignKeyWithMeta)
    })

    it('gets an existing key', async () => {
      const key = await keystore.getKeyByPath('ZZZ', SignKeyWithMeta)
      expect(key.publicKey.getLength()).toEqual(32)
      expect(key.secretKey.getLength()).toEqual(64)
    })

    it('throws an error upon accessing a non-existant key', async () => {
      try {
        await keystore.getKeyByPath('ZZZZ', SignKeyWithMeta)
      } catch (e) {
        assert(true)
      }
    })

    it('throws an error upon not receiving an ID', async () => {
      try {
        await keystore.getKeyByPath(undefined, undefined)
      } catch (e) {
        assert(true)
      }
    })

    it('throws an error accessing a closed store', async () => {
      try {
        await store.close()
        await keystore.getKeyByPath('ZZZ', SignKeyWithMeta)
      } catch (e) {
        assert(true)
      }
    })

    afterAll(async () => {
      // keystore.close()
    })
  })

  describe('getKeys', () => {
    let keystore: Keystore, aSignKey: KeyWithMeta, aBoxKey: KeyWithMeta, aBox2Key: KeyWithMeta, bSignKey: KeyWithMeta

    it('gets keys by group amd type', async () => {

      keystore = new Keystore(store)
      aSignKey = await keystore.createKey('asign', SignKeyWithMeta, 'group', { overwrite: true })
      aBoxKey = await keystore.createKey('abox', BoxKeyWithMeta, 'group', { overwrite: true })
      aBox2Key = await keystore.createKey('abox2', BoxKeyWithMeta, 'group', { overwrite: true })
      bSignKey = await keystore.createKey('bsign', SignKeyWithMeta, 'group2', { overwrite: true })


      const keysByGroup = await keystore.getKeys('group')
      expect(keysByGroup).toHaveLength(3);
      expect(keysByGroup.map(k => (k as (BoxKeyWithMeta | SignKeyWithMeta)).publicKey.toString('hex'))).toContainAllValues([aBoxKey, aBox2Key, aSignKey].map(k => (k as (BoxKeyWithMeta | SignKeyWithMeta)).publicKey.toString('hex')));
      const keysByType = await keystore.getKeys('group', BoxKeyWithMeta)
      expect(keysByType.map(k => (k as (BoxKeyWithMeta | SignKeyWithMeta)).publicKey.toString('hex'))).toContainAllValues([aBoxKey, aBox2Key].map(k => (k as (BoxKeyWithMeta | SignKeyWithMeta)).publicKey.toString('hex')));

    })
  })

  describe(SignKeyWithMeta, () => {
    let keystore: Keystore, key: SignKeyWithMeta, signingStore

    beforeAll(async () => {

      jest.setTimeout(10000)
      signingStore = await createStore(storagePath)
      keystore = new Keystore(signingStore) // 
      /*  await new Promise((resolve) => {
         setTimeout(() => {
           resolve(true);
         }, 3000);
       })*/
      /* 
      await keystore.close(); */
      /* const createdKey = await keystore.createKey('signing', SignKeyWithMeta, undefined, { overwrite: true })
      const y = deserialize(serialize(createdKey), KeyWithMeta); */
      key = await keystore.getKeyByPath('signing', SignKeyWithMeta)
      /* await keystore.close();  */ //
      const x = 123;
    })

    it('signs data', async () => {
      const signature = await Keystore.sign(Buffer.from('data data data'), key)
      expect(signature).toMatchSnapshot('signature');
    })

    it('throws an error if no key is passed', async () => {
      try {
        await Keystore.sign(Buffer.from('data data data'), null)
      } catch (e) {
        assert(true)
      }
    })

    it('throws an error if no data is passed', async () => {
      try {
        await Keystore.sign(null, key)
      } catch (e) {
        assert(true)
      }
    })

    afterAll(async () => {
      signingStore.close()
    })
  })

  describe('verify', () => {
    jest.setTimeout(5000)
    let keystore: Keystore, signingStore, publicKey: Ed25519PublicKey, key: SignKeyWithMeta

    beforeAll(async () => {
      signingStore = await createStore(storagePath)
      keystore = new Keystore(signingStore)
      key = await keystore.getKeyByPath('signing', SignKeyWithMeta)
      publicKey = key.publicKey
    })

    it('verifies content', async () => {
      const signature = '4FAJrjPESeuDK5AhhHVmrCdVbb6gTqczxex1CydLfCHH4kP4CBmoMLfH6UxLF2UmPkisNMU15RVHo63NbWiNvyyb2f4h8x5cKWtQrHY3mUL'
      try {
        const verified = await Keystore.verify(Buffer.from(signature), publicKey, Buffer.from('data data data'))
        expect(verified).toEqual(true)
      } catch (error) {
        const x = 123;
      }
    })

    it('verifies content with cache', async () => {
      const data = new Uint8Array(Buffer.from('data'.repeat(1024 * 1024)))
      const sig = await Keystore.sign(Buffer.from(data), key)
      const startTime = new Date().getTime()
      await Keystore.verify(sig, publicKey, Buffer.from(data))
      const first = new Date().getTime()
      await Keystore.verify(sig, publicKey, Buffer.from(data))
      const after = new Date().getTime()
      console.log('First pass:', first - startTime, 'ms', 'Cached:', after - first, 'ms')
      assert.strictEqual(first - startTime > after - first, true)
    })

    it('does not verify content with bad signature', async () => {
      const signature = 'xxxxxx'
      const verified = await Keystore.verify((new Uint8Array(Buffer.from(signature))), publicKey, Buffer.from('data data data'))
      expect(verified).toEqual(false)
    })

    afterAll(async () => {
      signingStore.close()
    })
  })

  describe('open', () => {
    let keystore: Keystore, signingStore

    beforeEach(async () => {
      signingStore = await createStore(storagePath)
      keystore = new Keystore(signingStore)
      signingStore.close()
    })

    it('closes then open', async () => {
      await waitFor(() => signingStore.status === 'closed');
      await keystore.openStore()
      expect(signingStore.status).toEqual('open')
    })

    it('fails when no store', async () => {
      let error = false
      try {
        keystore._store = undefined
        await keystore.openStore()
      } catch (e) {
        error = e.message
      }
      expect(error).toEqual('Keystore: No store found to open')
    })

    afterEach(async () => {
      signingStore.close()
    })
  })

  describe('encryption', () => {
    describe(BoxKeyWithMeta, () => {
      let keystore: Keystore, keyA: BoxKeyWithMeta, keyB: BoxKeyWithMeta, encryptStore

      beforeAll(async () => {
        encryptStore = await createStore(storagePath)
        keystore = new Keystore(encryptStore) // 

        await keystore.createKey('box-a', BoxKeyWithMeta);
        await keystore.createKey('box-b', BoxKeyWithMeta);
        keyA = await keystore.getKeyByPath('box-a', BoxKeyWithMeta)
        keyB = await keystore.getKeyByPath('box-b', BoxKeyWithMeta)

      })

      it('encrypts/decrypts', async () => {
        const data = new Uint8Array([1, 2, 3]);
        const encrypted = await keystore.encrypt(data, keyA, keyB.publicKey);
        const decrypted = await keystore.decrypt(encrypted, keyB, keyA.publicKey)
        assert.deepStrictEqual(data, decrypted);
      })


      afterAll(async () => {
        encryptStore.close()
      })
    })

  })

})