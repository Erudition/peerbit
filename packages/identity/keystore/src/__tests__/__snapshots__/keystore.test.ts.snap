// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`keystore SignKeyWithMeta signs data: signature 1`] = `
Uint8Array [
  44,
  124,
  192,
  165,
  144,
  131,
  28,
  203,
  80,
  254,
  104,
  109,
  85,
  68,
  167,
  227,
  146,
  52,
  54,
  237,
  101,
  248,
  191,
  179,
  23,
  251,
  90,
  131,
  0,
  6,
  15,
  182,
  71,
  131,
  153,
  198,
  238,
  242,
  201,
  74,
  184,
  130,
  34,
  250,
  254,
  15,
  116,
  150,
  195,
  128,
  104,
  45,
  214,
  129,
  70,
  30,
  157,
  139,
  140,
  19,
  16,
  189,
  191,
  1,
  100,
  97,
  116,
  97,
  32,
  100,
  97,
  116,
  97,
  32,
  100,
  97,
  116,
  97,
]
`;
